File: src/main/scala/li/cil/oc/common/tileentity/traits/PowerBalancer.scala
package li.cil.oc.common.tileentity.traits

import li.cil.oc.Settings
import li.cil.oc.api.network.Connector
import li.cil.oc.api.network.SidedEnvironment
import net.minecraft.util.EnumFacing

trait PowerBalancer extends PowerInformation with SidedEnvironment with Tickable {
  var globalBuffer, globalBufferSize = 0.0

  protected def isConnected: Boolean

  override def updateEntity() {
    super.updateEntity()
    if (isServer && isConnected && getWorld.getTotalWorldTime % Settings.get.tickFrequency == 0) {
      val nodes = connectors
      def network(connector: Connector) = if (connector != null && connector.network != null) connector.network else this
      // Yeeeeah, so that just happened... it's not a beauty, but it works. This
      // is necessary because power in networks can be updated asynchronously,
      // i.e. in separate threads (e.g. to allow screens to consume energy when
      // they change, which usually happens in a computers executor thread).
      // This multi-lock only happens in the main server thread, though, so we
      // don't have to fear deadlocks. I think.
      network(nodes(0)).synchronized {
        network(nodes(1)).synchronized {
          network(nodes(2)).synchronized {
            network(nodes(3)).synchronized {
              network(nodes(4)).synchronized {
                network(nodes(5)).synchronized {
                  val (sumBuffer, sumSize) = distribute()
                  if (sumSize > 0) {
                    val ratio = sumBuffer / sumSize
                    for (node <- connectors if isPrimary(node)) {
                      node.changeBuffer(node.globalBufferSize * ratio - node.globalBuffer)
                    }
                  }
                  globalBuffer = sumBuffer
                  globalBufferSize = sumSize
                }
              }
            }
          }
        }
      }
      updatePowerInformation()
    }
  }

  protected def distribute() = {
    var sumBuffer, sumSize = 0.0
    for (node <- connectors if isPrimary(node)) {
      sumBuffer += node.globalBuffer
      sumSize += node.globalBufferSize
    }
    (sumBuffer, sumSize)
  }

  private def connectors = EnumFacing.values.view.map(sidedNode(_) match {
    case connector: Connector => connector
    case _ => null
  })

  private def isPrimary(connector: Connector) = {
    val nodes = connectors
    connector != null && nodes(nodes.indexWhere(node => node != null && node.network == connector.network)) == connector
  }
}

File: src/main/scala/li/cil/oc/common/tileentity/traits/PowerBalancer.java
package li.cil.oc.common.tileentity.traits;

import li.cil.oc.Settings;
import li.cil.oc.api.network.Connector;
import li.cil.oc.api.network.EnvironmentHost;
import li.cil.oc.api.network.Network;
import net.minecraft.util.math.MathHelper;
import net.minecraft.world.World;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;

public final class PowerBalancer {
    public interface PowerBalancerHost extends EnvironmentHost {
        Iterable<Connector> getConnectorsToBalance();
    }

    private final PowerBalancerHost host;

    public PowerBalancer(final PowerBalancerHost host) {
        this.host = host;
    }

    public void balance() {
        final World world = host.getHostWorld();

        if (world.getTotalWorldTime() % Settings.get().tickFrequency() != 0) {
            return;
        }

        final Iterable<Connector> connectors = host.getConnectorsToBalance();

        final Set<Network> networks = new HashSet<>();
        for (final Connector connector : connectors) {
            final Network network = connector.getNetwork();
            if (network == null) {
                continue;
            }
            networks.add(network);
        }

        forEachLock(networks, ReentrantLock::lock);

        double sum = 0;
        double sizeSum = 0;
        for (final Network network : networks) {
            sum += network.getGlobalBuffer();
            sizeSum += network.getGlobalBufferSize();
        }

        final double fillRatio = MathHelper.clamp(sum / sizeSum, 0, 1);
        for (final Network network : networks) {
            network.changeBuffer(network.getGlobalBufferSize() * fillRatio - network.getGlobalBuffer());
        }

        forEachLock(networks, ReentrantLock::unlock);
    }

    private static void forEachLock(final Iterable<Network> networks, final Consumer<ReentrantLock> consumer) {
        for (final Network network : networks) {
            if (network != null) {
                if (network instanceof NetworkImpl) {
                    final NetworkImlp networkImpl = (NetworkImpl) network;
                    final ReentrantLock lock = networkImpl.getLock();
                    consumer.accept(lock);
                }
            }
        }
    }
}
